### ISSUE TITLE
rebase-popup offers reword but not fixup

### ISSUE THREAD

**Issue by DamienCassou** [magit/magit#2895]
The `magit-rebase-popup` offers the very nice `reword` and `edit` functions, but `fixup` is missing. It would be useful to rapidly fixup a commit when browsing the log. I can have a look and send a PR if you are interested.

**Comment by npostavs** [comment]
There is "fixup" and "Instant Fixup" on the commit popup. Does that do what you're thinking of?

**Comment by DamienCassou** [comment]
Noam Postavsky <notifications@github.com> writes:
> There is "fixup" and "Instant Fixup" on the commit popup. Does that do
> what you're thinking of?

no because I would like to fixup an old commit (from the log
buffer). The commit popup only allows fixing up the commit currently
being created.

-- 
Damien Cassou
http://damiencassou.seasidehosting.st

"Success is the ability to go from one failure to another without
losing enthusiasm." --Winston Churchill

**Comment by tarsius** [comment]
> The commit popup only allows fixing up the commit currently being created.

No, it allows fixing up any commit (reachable from HEAD). The fixup/squash commands are in the commit popup, not the rebase popup, because the first thing they do is to create a *new* commit (which becomes the new HEAD). Only after that has happened `git rebase` is used to move the changes to where they belong. If you use the "instant" variant, then that happens right away.

In the commit popup, only the first column is for creating a new commit in its own right, the second column is about manipulating HEAD, and columns tree and four are for editing older commits.

**Comment by tarsius** [comment]
I just realized that you responded by email. So I should probably tell you that I edited the previous rely a bit and you should read it in a browser, not your email program, so that you get to see the latest edits.

**Comment by tarsius** [comment]
Let me know if anything is still unclear after reading https://magit.vc/manual/magit/Initiating-a-commit.html.

**Comment by DamienCassou** [comment]
Thanks for you explanation @tarsius. Nevertheless, it's still unclear for me after reading both the documentation and your additional info. Imagine I have a clean working directory and 4 commits in the log with hashes a, b, c and d (d being the HEAD). I'm in the log buffer and want to fixup c to b. If I press `c f` to fixup, I get `user-error: Nothing staged (or unstaged)`.

**Comment by tarsius** [comment]
Ah, I didn't understand that it is this you are after. And yes, that's not really covered in the manual because it works exactly like with plain git and I don|'t have the time to duplicate all documentation.

Move to `b` and press <kbd>r i</kbd>. In the buffer that appears move to `c` and press <kbd>f</kbd> followed by <kbd>C-c C-c</kbd>.

**Comment by DamienCassou** [comment]
@tarsius now that we understand each other, could you please read the issue description again and tell me if it makes sense? Please let me rephrase it in case it becomes clearer. The `rebase interactive` (<kbd>r i</kbd>) tool is great when multiple changes have to be done. What I'm now looking after is a fast way to do something simple: from the log buffer, I want to fixup any commit to any other commit. The `rebase` popup already has `edit` and `reword` features that are faster than rebase interactive. I'm just asking if it would be possible to add `fixup` next to them.

**Comment by tarsius** [comment]
This could implemented using [the selection](https://magit.vc/manual/magit/Section-selection.html)*, i.e. like <kbd>d d</kbd> compares the commits at the edges of the selection, <kbd>r F/S</kbd> could fixup/squash one of the edge commits into the other.

(*) Note to self: this has to be documented in more depth in "Interface concepts".

**Comment by Silex** [comment]
Both `edit` and `reword` are actions that do not require any more input: you just edit or reword the selected commit.

`fixup` would require you to give one more information (to which commit you want it to fixup'ed).

I think in the meantime `r i` is really sufficient because it's not often that you decide to "fixup" a commit *after* it was created. You usually decide to make a `fixup` commit *at commit time*, and there `c f` or `c F` are perfect.

That said, why not make that feature one day if it's not too much work... would reuse the selection mechanism from `c f`.

**Comment by tarsius** [comment]
I've had another look and have decided not to implement this.

The proposed interfaces were:

> This could implemented using [the selection](https://magit.vc/manual/magit/Section-selection.html)*, i.e. like <kbd>d d</kbd> compares the commits at the edges of the selection, <kbd>r F/S</kbd> could fixup/squash one of the edge commits into the other.

Given this interface it would make more sense if all the commits in the range were squashed into one commit, instead of the commit on one end into the commit at the other end.

> reusing the selection mechanism from `c f` (by pressing `.`)

Let's assume point is on "Unpushed to feature-branch" and that `HEAD~1` is to be merged into `HEAD~3`. Using this proposed interface one would do this:

1. <kbd>C-n C-n</kbd> to go to temporary commit.
2. <kbd>r f</kbd> to initiate fixup rebase.
3. The "log selection" interface comes up.
4. <kbd>n n</kbd> to go to target commit.
5. <kbd>.</kbd> to select it and at the same time start the rebase.

Without this interface, using the generic interface, one would instead do this.

1. <kbd>r i</kbd> to initiate interactive rebase.
2. The usual "give details for rebase" interface comes up.
3. <kbd>n n</kbd> to go to temporary commit.
4. <kbd>M-n M-n</kbd> to move it right after (above) the target commit.
5. <kbd>f</kbd> to mark it as fixup commit.
6. <kbd>C-c C-c</kbd> to start the rebase.

This can be varied a bit. E.g. one could first move to the temporary commit and only then initiate the rebase. One could also first mark the temporary commit and then move it to the right place.

The effort required in these two workflows is very similar. The same number of movement commands are required. The main difference is that <kbd>.</kbd> is one command, while <kbd>k</kbd> followed by <kbd>C-c C-c</kbd> is two and also requires pressing more keys.

I don't think implementing the suggested very slightly more efficient but less flexible (and somewhat inconsistent) interface is worth the effort required to implement it. I also don't think many users would use it.

**Comment by tarsius** [comment]
Okay. I acknowledge that turning an existing commit into a fixup for another commit and then combining them is a bit laborious using the generic interface. But I also agree with this:

> I think in the meantime `r i` is sufficient because it's not often that you decide to "fixup" a commit *after* it was created. You usually decide to make a `fixup` commit *at commit time*, and there `c f` or `c F` are perfect.

Unfortunately implementing something better would be a lot of work. Because of that and because I think very few people would use this feature, I am not going to implement it. But if someone else implemented it, then I would most likely add it.

**Comment by DamienCassou** [comment]
Thank you for considering it. When I needed this feature last year, I needed it 10 times per hour during 2 days (huge rewrite of old feature branch). I haven't had the need for it since then (old features branches are bad).

**Comment by tarsius** [comment]
:wine_glass:
